Add trackball-like options for terrain interaction mode

adeak opened the pull request:
It was raised in https://github.com/pyvista/pyvista/issues/1022 that zooming should be enabled using the mouse wheel for terrain interaction mode, and in general more customizability would be welcome.

This PR tackles the easy part of the feature request: by passing `mouse_wheel_zoom=True` we can request zooming with the mouse wheel:

```py
import pyvista as pv
from pyvista import examples

mesh = examples.download_st_helens().warp_by_scalar()

plotter = pv.Plotter()
plotter.add_mesh(mesh)
plotter.enable_terrain_style(mouse_wheel_zoom=True)
plotter.show()
```

As far as I can tell the default right-click drag zooming should keep working too even if mouse-wheel zooming is requested ~(although I haven't figured out how to make that work on my touchpad :laughing:)~. Yup, it still works (note to self: double-fingered touchscreen click for zoom, triple-fingered touchscreen click for pan).

I've decided to add this mode as non-default for backwards compatibility, but since the original zooming behaviour is still present then perhaps we can just always add the mouse wheel zoom as an additional feature. Not sure if this could be disruptive. I checked that [the VTK interactor style doesn't seem to define an event for mouse wheel movements](https://github.com/Kitware/VTK/blob/master/Interaction/Style/vtkInteractorStyleTerrain.cxx), so we'd probably not be stepping on any toes.

I'm thinking perhaps I should also add the option to pan the same way that the default trackball style does it with shift + mouse move...

adeak commented:
I've added shift plus left mouse button panning. I'm 80% sure I'm not doing something dumb... so thorough reviews will be appreciated :)

I'm becoming certain that these optional features should actually be additional features by default. Question is whether it might break something downstream. How likely is that downstream overrides mouse scrolling or shift + left mouse button to something surprising?

MatthewFlamm commented:
How would this fit in with this comment https://github.com/pyvista/pyvista/issues/1022#issuecomment-729787296 for providing a way to create a custom interactor style?  If that could be done by overriding pieces of an existing style, that would be ideal in my view.  None of the other styles allow custom overrides, which might get messy to support one at a time like this PR.

adeak commented:
> If that could be done by overriding pieces of an existing style, that would be ideal in my view. 

Yes, _ideally_ we can enable full customization. _Pragmatically_ there's a common need for an interaction style that's "the default except up is always up, like in matplotlib or MATLAB". I have this need myself. And it's not clear to me from an API design nor an implementation perspective how we could pull off arbitrary customization.

Given the current implementation of wrapping interaction styles it's the least effort to patch the terrain style rather than creating a new style from scratch or implementing full customizability.

> None of the other styles allow custom overrides, which might get messy to support one at a time like this PR.

In what way do you think it would be "messy"? This feature doesn't even have to be opt-in, and even if it is, we could always reimplement the functionality using the hypothetical custom override system of the future. Whenever that happens.

MatthewFlamm commented:
These are all fair points.  I don't have a strong stake in the ground.  If there is a good reason for this style modification, then I think it is OK.

One way that it could get messy is that some of these styles have built-in modes that can be customized in vtk, so there could be a mix of built-in modifications and custom modifications.

https://vtk.org/doc/nightly/html/classvtkInteractorStyleImage.html#details

adeak commented:
Thanks for the link. The way I would approach such a built-in customization option is through a keyword argument to the corresponding `enable_*()` method, i.e. the same way that I'm trying to generalize here. So this mostly seems like a potential counterargument to general customization. Or at least an important caveat when design decisions are made to support general customization.

And your link prompted me to check the same for Terrain (I should've done that in the first place!): https://vtk.org/doc/nightly/html/classvtkInteractorStyleTerrain.html#details

Turns out shift+mouse is already meaningful:

> Therefore, diagonal motion results in a combination of azimuth and elevation. (If the shift key is held during motion, then only one of elevation or azimuth is invoked, depending on the whether the mouse motion is primarily horizontal or vertical.) Middle mouse button pans the camera across the scene (again the shift key has a similar effect on limiting the motion to the vertical or horizontal direction.

And indeed that's what the default does. But the effect is subtle enough that I didn't notice during my experiments. You have to intentionally use broad diagonal movements to make this obvious.

This tells me that the new feature should definitely be opt-in, because users might rely on the current shift+move behaviour. Thanks.

akaszynski commented on tests/plotting/test_plotting.py:
```diff
@@ -227,22 +227,22 @@ def test_plot_invalid_style(sphere):
 
 @skip_no_plotting
-def test_interactor_style(sphere):
+@pytest.mark.parametrize('interaction, kwargs', [
+    ('trackball', {}),
+    ('trackball_actor', {}),
+    ('image', {}),
+    ('joystick', {}),
+    ('zoom', {}),
+    ('terrain', {}),
+    ('terrain', {'mouse_wheel_zooms': True, 'shift_pans': True}),
+    ('rubber_band', {}),
+    ('rubber_band_2d', {}),
+])
+def test_interactor_style(sphere, interaction, kwargs):
```
Agree with the parameterization.

akaszynski commented on pyvista/plotting/render_window_interactor.py:
```diff
@@ -233,18 +233,64 @@ def enable_zoom_style(self):
+            shift + left mouse button rotates the view restricted to only
+            horizontal or vertical movements, and panning is done holding
+            down the middle mouse button.
+
```
Single request is to add an Example section to this.  One of the best things about `numpy` documentation is its examples.  Half the time I don't even read the parameter descriptions and just look to see how its used.

adeak replied:
Good idea. Let me try to do that for every style.

adeak replied:
Added in https://github.com/pyvista/pyvista/pull/1376/commits/377c61fd2e2449fb153729c025a51684f33d5dc5

akaszynski reviewed (approved):
Approved with one request to add an example section.  It's basic, but we really need to start adding examples to our methods.

adeak commented:
I've added examples and list of controls to every interaction style, and added the missing Joystick Actor style.

I suspect some of these styles don't make complete sense in pyvista right now. For instance the [Image style](https://vtk.org/doc/nightly/html/classvtkInteractorStyleImage.html#details) seems to make sense with `vtkImageActors`, which only seem to appear in pyvista in the undocumented `BackgroundRenderer` class. [Rubber Band Pick](https://vtk.org/doc/nightly/html/classvtkInteractorStyleRubberBandPick.html#details) seems to need an "attached picker"? I don't know VTK and pyvista enough to be sure, but these are suspect.

akaszynski commented:
> I've added examples and list of controls to every interaction style, and added the missing Joystick Actor style.

Thanks!  Docs look good.

> I suspect some of these styles don't make complete sense in pyvista right now.

Perhaps not, but we can leave that for a later PR to get full docs and implementation detail right.

You're good to merge.
